cmd/ssh.go

```diff
@@ -284,3 +456,88 @@ func (cmd *SSHCmd) startServices(ctx context.Context, devPodConfig *config.Confi
 		}
 	}
 }
+
+// setupGPGAgent will forward a local gpg-agent into the remote container
+// this works by using cmd/agent/workspace/setup_gpg
+func (cmd *SSHCmd) setupGPGAgent(
+	ctx context.Context,
+	containerClient *ssh.Client,
+	log log.Logger,
+) error {
+	writer := log.ErrorStreamOnly().Writer(logrus.InfoLevel, false)
+	defer writer.Close()
+
+	log.Debugf("gpg: exporting gpg public key from host")
+
+	// Read the user's public keys and ownertrust from GPG.
+	// These commands are executed LOCALLY, the output will be imported by the remote gpg
+	pubKeyExport, err := exec.Command("gpg", "--armor", "--export").Output()
+	if err != nil {
+		return fmt.Errorf("export local public keys from GPG: %w", err)
+	}
+
+	log.Debugf("gpg: exporting gpg owner trust from host")
+
+	ownerTrustExport, err := exec.Command("gpg", "--export-ownertrust").Output()
+	if err != nil {
+		return fmt.Errorf("export local ownertrust from GPG: %w", err)
+	}
+
+	log.Debugf("gpg: detecting gpg-agent socket path on host")
+	// Detect local agent extra socket, this will be forwarded to the remote and
+	// symlinked in multiple paths
+	gpgExtraSocketBytes, err := exec.Command("gpgconf", []string{"--list-dir", "agent-extra-socket"}...).
+		Output()
+	if err != nil {
+		return err
+	}
+
+	gpgExtraSocketPath := strings.TrimSpace(string(gpgExtraSocketBytes))
+	log.Debugf("gpg: detected gpg-agent socket path %s", gpgExtraSocketPath)
+
+	log.Debugf("ssh: starting reverse forwarding socket %s", gpgExtraSocketPath)
+	cmd.ReverseForwardPorts = append(cmd.ReverseForwardPorts, gpgExtraSocketPath)
+
+	go func() {
+		err := cmd.reverseForwardPorts(ctx, containerClient, log)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}()
+
+	pubKeyArgument := base64.StdEncoding.EncodeToString(pubKeyExport)
+	ownerTrustArgument := base64.StdEncoding.EncodeToString(ownerTrustExport)
+
+	// Now we forward the agent socket to the remote, and setup remote gpg to use it
+	// fix eventual permissions and so on
+	forwardAgent := []string{
+		agent.ContainerDevPodHelperLocation,
+		"agent",
+		"workspace",
+		"setup-gpg",
+		"--publickey",
+		pubKeyArgument,
+		"--ownertrust",
+		ownerTrustArgument,
+		"--socketpath",
+		gpgExtraSocketPath,
+	}
+
+	if log.GetLevel() == logrus.DebugLevel {
+		forwardAgent = append(forwardAgent, "--debug")
+	}
+
+	log.Debugf(
+		"gpg: start reverse forward of gpg-agent socket %s, keeping connection open",
+		gpgExtraSocketPath,
+	)
+
+	command := strings.Join(forwardAgent, " ")
+
+	if cmd.User != "" && cmd.User != "root" {
+		command = fmt.Sprintf("su -c \"%s\" '%s'", command, cmd.User)
+	}
+
+	// We use start to keep this connection alive in background (hence the use of sleep infinity)
```

Comment:
This comment seems to be somewhat detached from the code, could you either elaborate or rewrite pls?

Reply:
Ah it's a leftover from a previous implementation, I'll remove it

Reply:
Removed